File: ios/Classes/WebPaymentsHandler.swift
//
//  WebPaymentsHandler.swift
//  Pods
//
//  Created by Amit on 14/12/2025.
//

import Flutter
import Foundation
import PayPal

/**
 * Handles PayPal Web Payment flow for the Flutter iOS platform.
 */
class WebPaymentsHandler: NSObject, PaypalWebPaymentHostApi {
    private let paypalConfig: PaypalPaymentConfig
    private let messenger: FlutterBinaryMessenger
    
    init(paypalConfig: PaypalPaymentConfig, messenger: FlutterBinaryMessenger) {
        self.paypalConfig = paypalConfig
        self.messenger = messenger
        super.init()
        
        // Setup the host API on the messenger
        PaypalWebPaymentHostApiSetup.setUp(binaryMessenger: messenger, api: self)
    }
    
    
    /// Starts the PayPal Web Payment flow for the given order and funding source.
    /// - Parameters:
    ///   - orderId: The PayPal order ID to be processed.
    ///   - fundingSource: The selected funding source (e.g., "paypal", "card").
    /// - Throws: `PigeonError` if the SDK is not initialized.
    func initiatePaymentRequest(
        orderId: String,
        fundingSource: String
    ) throws {
        
        // MARK: STEP 1 — Create & Register Checkout Result Listener
        // This listener receives success/failure callbacks from the SDK.
        // It is registered through the event stream handler so results can be
        // forwarded back to Dart/Flutter.
        let listener = WebPaymentRequestResultEventListener()
        
        // Register the event listener with Flutter’s binary messenger.
        WebPaymentRequestResultEventListener.register(
            with: messenger,
            streamHandler: listener
        )
        
        // MARK: STEP 2 — Validate that the SDK has been initialized
        // Validate that the SDK has been initialized and a valid client ID exists.
        // If not available, checkout cannot proceed.
        guard let clientId = paypalConfig.clientId else {
            listener.onError("Plugin not initialized. Please initialize the SDK first.")
            return
        }
        
        // MARK: STEP 3 — Create PayPal Core Configuration
        // This configuration object defines the client ID and environment
        // (sandbox/live) used by PayPal's SDK to operate the checkout flow.
        let config = CoreConfig(
            clientID: clientId,
            environment: paypalConfig.toPaypalEnvironment()
        )
        
        // MARK: STEP 4 — Initialize Web Checkout Client
        // The checkout client is responsible for launching and managing
        // the PayPal web-based authorization and payment process.
        let checkoutClient = PayPalWebCheckoutClient(config: config)
        
        // MARK: STEP 5 — Build Checkout Request
        // This request includes the order ID and selected funding source
        // and is passed to the PayPal checkout client to initiate the flow.
        let request = PayPalWebCheckoutRequest(
            orderID: orderId,
            fundingSource: fundingSource.toPayPalWebCheckoutFundingSource()
        )
        
        // MARK: STEP 6 — Start Checkout Flow
        // The `start` method triggers the PayPal authorization experience.
        // The callback reports initial validation or technical failures.
        checkoutClient.start(request: request, completion: { result in
            switch result {
                case .failure(let error):
                    // Handles any failure occurring before the checkout UI loads
                    // (e.g., invalid request, network failure).
                    listener.onFailure(
                        orderId: orderId,
                        reason: error.errorDescription ?? "Unknown authorization error",
                        code: error.code ?? 0,
                        correlationId: nil
                    )
                    
                case .success(let webCheckoutResult):
                    // Checkout launched successfully and final approval completed.
                    // The result contains order and payer IDs.
                    listener.onSuccess(
                        orderId: webCheckoutResult.orderID,
                        payerId: webCheckoutResult.payerID
                    )
            }
        })
    }
    
    /**
     * Event stream handler that bridges PayPal Web Payment Request results from native iOS to Flutter.
     * This listener captures the Flutter event sink when Dart starts listening and sends exactly
     * one terminal event (success, failure, canceled, or error). After emitting, the stream is
     * closed and the sink is cleared.
     */
    private class WebPaymentRequestResultEventListener: PaypalWebPaymentRequestResultEventStreamHandler {
        
        // The active Flutter event sink used to emit `PaypalWebPaymentRequestResultEvent` values back to Dart.
        private var eventSink: PigeonEventSink<PaypalWebPaymentRequestResultEvent>?
        
        // Called when the Dart side begins listening to the event stream.
        // Stores the provided sink so native can send result events back to Flutter.
        override func onListen(
            withArguments arguments: Any?,
            sink: PigeonEventSink<PaypalWebPaymentRequestResultEvent>
        ) {
            self.eventSink = sink
        }
        
        // Called when Dart cancels its subscription.
        // Releases the sink reference to prevent further emissions.
        override func onCancel(withArguments arguments: Any?) {
            self.eventSink = nil
        }
        
        // Ends the event stream by notifying Flutter and clearing the sink.
        // Called internally after sending any terminal event.
        private func end() {
            eventSink?.endOfStream()
            self.eventSink = nil
        }
        
        // Emits a successful checkout result and closes the stream.
        func onSuccess(orderId: String?, payerId: String?) {
            let event = PaypalWebPaymentRequestSuccessResultEvent(
                orderId: orderId,
                payerId: payerId
            )
            eventSink?.success(event)
            end()
        }
        
        // Emits a failure result and closes the stream.
        func onFailure(
            orderId: String?,
            reason: String,
            code: Int,
            correlationId: String?
        ) {
            let event = PaypalWebPaymentRequestFailureResultEvent(
                orderId: orderId,
                reason: reason,
                code: Int64(code),
                correlationId: correlationId
            )
            eventSink?.success(event)
            end()
        }
        
        // Emits a cancellation result and closes the stream.
        func onCanceled(orderId: String?) {
            let event = PaypalWebPaymentRequestCanceledResultEvent(orderId: orderId)
            eventSink?.success(event)
            end()
        }
        
        // Emits a generic error event and closes the stream.
        func onError(_ error: String) {
            let event = PaypalWebPaymentRequestErrorResultEvent(error: error)
            eventSink?.success(event)
            end()
        }
    }
    
    
}

// Helper extension function to convert string funding source to
// Paypal webcheckout funding source
extension String {
    /**
     Converts a textual funding source into a `PayPalWebCheckoutFundingSource`.
     
     - Returns: The matching `PayPalWebCheckoutFundingSource` value. If the string
     does not match a known funding source, `.paypal` is returned by default.
     
     Supported inputs (case-sensitive):
     - "paypal"     -> `.paypal`
     - "credit"     -> `.paypalCredit`
     - "payLater"   -> `.paylater`
     
     Notes:
     - This helper is intentionally conservative and case-sensitive to avoid
     accidental mappings. Normalize your input (e.g., lowercasing) upstream if
     needed.
     - The default branch returns `.paypal` to provide a safe fallback that
     continues the checkout flow with the most common funding source.
     */
    func toPayPalWebCheckoutFundingSource() -> PayPalWebCheckoutFundingSource {
        switch self {
            case "paypal":
                return .paypal
            case "credit":
                return .paypalCredit
            case "payLater":
                return .paylater
            default:
                return .paypal
        }
    }
}

File: android/src/main/kotlin/com/itheamc/paypal_payment_flutter/WebPaymentsHandler.kt
package com.itheamc.paypal_payment_flutter

import FlutterError
import PaypalWebPaymentHostApi
import PaypalWebPaymentRequestCanceledResultEvent
import PaypalWebPaymentRequestErrorResultEvent
import PaypalWebPaymentRequestFailureResultEvent
import PaypalWebPaymentRequestResultEvent
import PaypalWebPaymentRequestResultEventStreamHandler
import PaypalWebPaymentRequestSuccessResultEvent
import PigeonEventSink
import android.app.Activity
import android.content.Intent
import com.paypal.android.corepayments.CoreConfig
import com.paypal.android.paypalwebpayments.PayPalPresentAuthChallengeResult
import com.paypal.android.paypalwebpayments.PayPalWebCheckoutClient
import com.paypal.android.paypalwebpayments.PayPalWebCheckoutFinishStartResult
import com.paypal.android.paypalwebpayments.PayPalWebCheckoutFundingSource
import com.paypal.android.paypalwebpayments.PayPalWebCheckoutRequest
import io.flutter.embedding.engine.plugins.FlutterPlugin
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

/**
 * Handles PayPal web checkout flow.
 *
 * @property getConfig The callback to get the paypal payment configuration
 * @property getActivity The callback to get the current activity.
 * @property getPluginBinding The callback to get the current plugin binding.
 */
class WebPaymentsHandler(
    private val getConfig: () -> PaypalPaymentConfig,
    private val getActivity: () -> Activity?,
    private val getPluginBinding: () -> FlutterPlugin.FlutterPluginBinding?,
) : PaypalWebPaymentHostApi {

    /**
     * The client for handling PayPal web checkout.
     */
    private var checkoutClient: PayPalWebCheckoutClient? = null

    /**
     * The listener for checkout result events.
     */
    private var listener: PaypalWebPaymentRequestResultEventListener? = null

    /**
     * The flag to check if the checkout on intent is called or not
     */
    private var isCheckoutOnIntentCalled = false

    /**
     * Currently processing order id
     */
    private var currentlyProcessingOrderId: String? = null

    init {
        getPluginBinding()?.let {
            PaypalWebPaymentHostApi.setUp(it.binaryMessenger, this)
        }
    }

    /**
     * Initiates the PayPal web checkout.
     *
     * @param orderId The ID of the order to be processed.
     * @param fundingSource The funding source for the payment.
     * @throws FlutterError if the activity, binding, or PayPal configuration is null.
     */
    override fun initiatePaymentRequest(
        orderId: String,
        fundingSource: String
    ) {
        // STEP 1: Make checkout intent called variable false and set the
        // currently processing order id
        isCheckoutOnIntentCalled = false
        currentlyProcessingOrderId = orderId

        // STEP 2: Initialize & register Result Listener
        listener = PaypalWebPaymentRequestResultEventListener()

        PaypalWebPaymentRequestResultEventStreamHandler.register(
            getPluginBinding()!!.binaryMessenger,
            listener!!
        )

        // STEP 3: Creating Configuration Object for Paypal Web Payment
        val paypalConfig = getConfig()

        if (paypalConfig.clientId == null) {
            listener?.onError("SDK not initialized")
            listener = null
            currentlyProcessingOrderId = null
            return
        }

        val config = CoreConfig(
            paypalConfig.clientId!!,
            paypalConfig.toPaypalEnvironment()
        )

        // STEP 4: Initializing Checkout Client
        checkoutClient = PayPalWebCheckoutClient(
            context = getActivity()!!,
            configuration = config,
            urlScheme = "itheamc://paypal"
        )

        // STEP 5: Creating the web checkout request
        val request = PayPalWebCheckoutRequest(
            orderId = orderId,
            fundingSource = fundingSource.toPayPalWebCheckoutFundingSource()
        )

        // STEP 6: Starting Checkout Request
        checkoutClient?.start(
            activity = getActivity()!!,
            request = request,
            callback = { result ->
                when (result) {
                    is PayPalPresentAuthChallengeResult.Failure -> {
                        listener?.onFailure(
                            orderId = orderId,
                            reason = result.error.errorDescription,
                            code = result.error.code,
                            correlationId = result.error.correlationId
                        )
                    }

                    is PayPalPresentAuthChallengeResult.Success -> {
                        // Do nothing here
                    }
                }
            }
        )
    }

    /**
     * Handles the result from the PayPal web checkout flow.
     *
     * @param intent The intent containing the result data.
     * @return `true` if the intent was handled, `false` otherwise.
     */
    fun onNewIntent(intent: Intent): Boolean {

        // If the checkout client or currently processing order id is null, return false
        if (checkoutClient == null || currentlyProcessingOrderId == null) return false;

        // Else if the intent scheme is "itheamc", go ahead and finish start checkout
        if (intent.scheme == "itheamc") {
            // STEP 7: Make checkout intent called variable true
            isCheckoutOnIntentCalled = true

            // STEP 8: Finish Start Checkout
            when (val result = checkoutClient?.finishStart(intent)) {
                is PayPalWebCheckoutFinishStartResult.Canceled -> {
                    listener?.onCanceled(
                        orderId = result.orderId,
                    )
                }

                is PayPalWebCheckoutFinishStartResult.Failure -> {
                    listener?.onFailure(
                        orderId = result.orderId,
                        reason = result.error.errorDescription,
                        code = result.error.code,
                        correlationId = result.error.correlationId
                    )
                }

                PayPalWebCheckoutFinishStartResult.NoResult -> {
                    listener?.onSuccess(
                        orderId = intent.data?.getQueryParameter("token"),
                        payerId = intent.data?.getQueryParameter("PayerID")
                    )
                }

                is PayPalWebCheckoutFinishStartResult.Success -> {
                    listener?.onSuccess(
                        orderId = result.orderId,
                        payerId = result.payerId
                    )
                }

                null -> {
                    listener?.onError("Something went wrong")
                }
            }

            // STEP 9: Reset client and listener
            checkoutClient = null
            listener = null
            currentlyProcessingOrderId = null
            return true
        }
        return false
    }

    /**
     * Handles the lifecycle event when the app is resumed.
     *
     * This method checks if the user has returned to the app without completing the
     * checkout process via the deep link (`onNewIntent`). This can happen, for example,
     * if the user manually navigates back to the app or closes the browser.
     *
     * If `onNewIntent` was not called and a checkout session is active, it waits for a
     * brief moment (2 seconds) to ensure it's not a false positive, then cancels the
     * checkout process, sends a cancellation event, and cleans up the associated resources.
     *
     * The early return `if (_isCheckoutOnIntentCalled) return` prevents this cancellation
     * logic from running if the checkout flow has already been handled by `onNewIntent`.
     */
    fun onResume() {
        if (checkoutClient != null && listener != null) {
            CoroutineScope(Dispatchers.Main).launch {
                delay(2000)

                // If the checkout flow has already been handled by onNewIntent, return early
                if (isCheckoutOnIntentCalled) return@launch

                // Cancel the checkout process and send a cancellation event
                listener?.onCanceled(currentlyProcessingOrderId)

                checkoutClient = null
                listener = null
                currentlyProcessingOrderId = null
            }
        }
    }

    /**
     * An event listener for PayPal web payment request results.
     */
    private class PaypalWebPaymentRequestResultEventListener :
        PaypalWebPaymentRequestResultEventStreamHandler() {
        private var eventSink: PigeonEventSink<PaypalWebPaymentRequestResultEvent>? = null

        override fun onListen(p0: Any?, sink: PigeonEventSink<PaypalWebPaymentRequestResultEvent>) {
            eventSink = sink
        }

        override fun onCancel(p0: Any?) {
            eventSink = null
        }

        /**
         * Handles a successful checkout.
         *
         * @param orderId The ID of the order.
         * @param payerId The ID of the payer.
         */
        fun onSuccess(
            orderId: String?,
            payerId: String?
        ) {
            eventSink?.success(
                PaypalWebPaymentRequestSuccessResultEvent(
                    orderId = orderId,
                    payerId = payerId
                )
            )

            end()
        }

        /**
         * Handles a failed checkout.
         *
         * @param orderId The ID of the order.
         * @param reason The reason for the failure.
         * @param code The error code.
         * @param correlationId The correlation ID for the request.
         */
        fun onFailure(
            orderId: String?,
            reason: String,
            code: Int,
            correlationId: String?
        ) {
            eventSink?.success(
                PaypalWebPaymentRequestFailureResultEvent(
                    orderId = orderId,
                    reason = reason,
                    code = code.toLong(),
                    correlationId = correlationId
                )
            )

            end()
        }

        /**
         * Handles a canceled checkout.
         *
         * @param orderId The ID of the order.
         */
        fun onCanceled(orderId: String?) {
            eventSink?.success(
                PaypalWebPaymentRequestCanceledResultEvent(
                    orderId = orderId
                )
            )

            end()
        }

        /**
         * Handles an error during checkout.
         *
         * @param error The error message.
         */
        fun onError(error: String) {
            eventSink?.success(
                PaypalWebPaymentRequestErrorResultEvent(
                    error = error
                )
            )

            end()
        }

        /**
         * Ends the event stream.
         */
        private fun end() {
            eventSink?.endOfStream()
            eventSink = null
        }
    }
}


/**
 * Converts a string to a [PayPalWebCheckoutFundingSource].
 *
 * @return The corresponding [PayPalWebCheckoutFundingSource], or [PayPalWebCheckoutFundingSource.PAYPAL] if the string is not recognized.
 */
private fun String.toPayPalWebCheckoutFundingSource(): PayPalWebCheckoutFundingSource {
    return when (this) {
        "paypal" -> PayPalWebCheckoutFundingSource.PAYPAL
        "credit" -> PayPalWebCheckoutFundingSource.PAYPAL_CREDIT
        "payLater" -> PayPalWebCheckoutFundingSource.PAY_LATER
        else -> PayPalWebCheckoutFundingSource.PAYPAL
    }
}